### PR TITLE
Feature/nette 3.0

### LINKED ISSUE #16: Verze pro Nette 3.0?
Prosím, kdy bude verze pro Nette 3.0?

### PR DESCRIPTION
**Pull request by laster04**
Closes #16 
update composer for nette 3.0 and beforeCompile update

### CODE REVIEW

**Comment by vojtechmares** [comment]
Hello, fist of all - thank you for the PR.

Please, can you add `Schema` for configuration? Since Nette 3, for defaults, we are using `nette/schema` package to create configuration schema for each package.

If you don't know about it, see https://github.com/contributte/elasticsearch/blob/master/src/DI/ElasticsearchExtension.php#L18, I think it will help you understand and write schema for this package.

Also when you are at it, please change method https://github.com/laster04/image-storage/blob/feature/nette-3.0/src/DI/ImageStorageExtension.php#L30 `setClass` to `setFactory` since `setClass` is deprecated and instead of passing a string of namespace address, use `use` statement and `::class`, like `$builder->setFactory(ExampleTarget::class)`.

And please, squash the three commits into one :smile:

**Comment by laster04** [comment]
@vojtamares 
Hi thx for review. I'am sorry this is my first PR ever :smile: . 
Please can you give me a hint how can i set up defaults? thx

**Comment by vojtechmares** [comment]
No worries :), we thank you for your PR and contribution.

An example of config schema can be found here: https://github.com/contributte/elasticsearch/blob/master/src/DI/ElasticsearchExtension.php#L18

Practically we are replacing the private property `$defaults` with a function function `getConfigSchema()` which returns a `Schema`, the schema itself is created with static methods from `Expect` class (`Nette/Schema/Expect`).

If you check the `use` imports at the top of the file and then the function you can get the gist I think. Or have a look into other repositories under [Contributte](https://github.com/contributte) organisation where Nette 3 is already supported.

**Comment by vojtamares** [comment]
No worries :), we thank you for your PR and contribution.

An example of config schema can be found here: https://github.com/contributte/elasticsearch/blob/master/src/DI/ElasticsearchExtension.php#L18

Practically we are replacing the private property `$defaults` with a function function `getConfigSchema()` which returns a `Schema`, the schema itself is created with static methods from `Expect` class (`Nette/Schema/Expect`).

If you check the `use` imports at the top of the file and then the function you can get the gist I think. Or have a look into other repositories under [Contributte](https://github.com/contributte) organisation where Nette 3 is already supported.

File: src/DI/ImageStorageExtension.php [hunk 1]
```diff
@@ -10,40 +10,49 @@ class ImageStorageExtension extends CompilerExtension
...
-		'friendly_url'       => false,
+		'friendly_url' => false,
 	];
 
-	public function loadConfiguration(): void
-	{
-		$config = $this->_getConfig();
+	public function getConfigSchema (): Nette\Schema\Schema {
+		return Nette\Schema\Expect::structure([
+			'data_path' => Nette\Schema\Expect::string(),
```
Comment (mabar) [inline src/DI/ImageStorageExtension.php]: `string()->required()`

Comment (mabar) [reply src/DI/ImageStorageExtension.php]: Or get `wwwDir` from `$this->getContainerBuilder()->parameters['wwwDir']` and add `parameters['wwwDir'] . '/../public/data'` as default value. I would prefer required.

File: src/DI/ImageStorageExtension.php [hunk 2]
```diff
@@ -10,40 +10,49 @@ class ImageStorageExtension extends CompilerExtension
...
+			'data_dir' => Nette\Schema\Expect::string(),
```
Comment (mabar) [inline src/DI/ImageStorageExtension.php]: Same as previous. `string()->required()` or `string('data')`

File: src/DI/ImageStorageExtension.php [hunk 3]
```diff
@@ -10,40 +10,49 @@ class ImageStorageExtension extends CompilerExtension
...
+			'algorithm_file' => Nette\Schema\Expect::string(),
```
Comment (mabar) [inline src/DI/ImageStorageExtension.php]: `string('sha1_file')`

File: src/DI/ImageStorageExtension.php [hunk 4]
```diff
@@ -10,40 +10,49 @@ class ImageStorageExtension extends CompilerExtension
...
-		'default_transform'  => 'fit',
+		'data_path' => '%wwwDir%/../public/data',
+		'data_dir' => 'data',
+		'algorithm_file' => 'sha1_file',
+		'algorithm_content' => 'sha1',
+		'quality' => 85,
+		'default_transform' => 'fit',
 		'noimage_identifier' => 'noimage/03/no-image.png',
-		'friendly_url'       => false,
+		'friendly_url' => false,
```
Comment (mabar) [inline src/DI/ImageStorageExtension.php]: You can remove whole $defaults, it is not used anymore as default are parts of schema (`'friendly_url' => Nette\Schema\Expect::bool(false)`) etc.

Comment (laster04) [reply src/DI/ImageStorageExtension.php]: Yea i see it now, sorry i did't look at Expect class definition :(

Comment (mabar) [reply src/DI/ImageStorageExtension.php]: It's not easy to migrate, schema works absolutelly different way. Feel free to contact me at [slack channel](https://pehapkari.herokuapp.com) of pehapkari.cz for further help

File: src/DI/ImageStorageExtension.php [hunk 5]
```diff
@@ -2,66 +2,49 @@
...
-		'algorithm_content'  => 'sha1',
-		'quality'            => 85,
-		'default_transform'  => 'fit',
-		'noimage_identifier' => 'noimage/03/no-image.png',
-		'friendly_url'       => false,
-	];
+	public function getConfigSchema(): Schema
+	{
+		return Expect::structure([
+			'data_path' => Expect::string('%wwwDir%/../public/data'),
```
Comment (mabar) [inline src/DI/ImageStorageExtension.php]: Schema does not automatically expand %parameter%, you need to do what I described ;)

**Comment by PavelJurasek** [comment]
Is there something to be done?

**Comment by PavelJurasek** [comment]
Bump

**Comment by paveljanda** [comment]
Cool, thanks!

**Pull request merged**